src/dinit.cc

```diff
@@ -696,6 +705,30 @@ int dinit_main(int argc, char **argv)
     close_control_socket();
     
     if (am_system_mgr) {
+        if (shutdown_type == shutdown_type_t::SOFTREBOOT) {
+            sync(); // Sync to minimise data loss if user elects to power off / hard reset
+            if (!opts.ignore_sys_args) {
+                char **argv_exec = (char **)malloc((argc + 2) * sizeof(char *));
```

Comment:
One of the main principles of Dinit is ["Robustness (including allocation robustness)"](https://github.com/davmac314/dinit/blob/0b5158328427778f1e7d28104bc8cca43aef572a/doc/COMPARISON#L27-L31). The `malloc()` call could fail and it's not handled here and can result in null dereference in further usage (see L713).

I think malloc failure should be handled here.

Reply:
I need a little time to consider the whole issue. I think it is probably better to filter arguments out before re-exec, or even to re-build a new command line from scratch, than it is to have a new argument which effectively filters out arguments at a different stage. Please allow me a little time to get back to this issue with a response once I've thought it through. @mobin-2008 is correct though that we cannot ignore `malloc` failures. It's also generally preferable to use `new` rather than `malloc`, note that this throws an exception on failure and that needs to be handled.

Reply:
Okay, I've now filtered the `single` argument in `argv`, and handled the exception in `new`.